File: Python/queue/queue.py
# Implementacao Classica de uma Fila(queue) Circular.


class Queue:
    def __init__(self, max_length):
        self.items = [0] * max_length
        self.max_length = max_length
        self.front = 0
        self.pointer = 0
        self.size = 0

    def enqueue(self, value):
        if self.pointer >= self.max_length:
            raise IndexError("Full queue")
        
        self.items[self.pointer] = value
        self.pointer = (self.pointer + 1) % self.max_length
        self.size += 1

    def dequeue(self):
        if self.is_empty():
            raise IndexError("Empty queue")
        
        value = self.items[self.front]
        self.items[self.front] = None
        self.front = (self.front + 1) % self.max_length
        self.size -= 1
        return value
    
    def peek(self):
        if self.is_empty():
            raise IndexError("Empty queue")
        
        return self.items[self.front]
    
    def is_empty(self):
        return self.size == 0
    
    def size(self):
        return self.size

File: Java/java-algorithms/src/Queue/queue.java
package Queue;

// Implementacao Classica de uma Fila(queue) Circular.

public class queue<T> {
    private final Object[] items;
    private final int capacity;
    private int front;
    private int pointer;
    private int size;

    public queue(int capacity) {
        this.capacity = capacity;
        items = new Object[capacity];
        front = 0;
        pointer = 0;
        size = 0;
    }

    public void enqueue(T value) {
        if (size >= capacity) {
            throw new IndexOutOfBoundsException("Queue is full");
        }
        items[pointer] = value;
        pointer = (pointer + 1) % capacity;
        size++;
    }

    public T dequeue() {
        if (size == 0) {
            throw new IndexOutOfBoundsException("Queue is empty");
        }
        T value = (T) items[front];
        items[front] = null;
        front = (front + 1) % capacity;
        size--;
        return value;
    }

    public T peek() {
        if (size == 0) {
            throw new IndexOutOfBoundsException("Queue is empty");
        }
        return (T) items[front];
    }

    public int size() {
        return size;
    }
}
